pkg/query-service/agentConf/manager.go

```diff
@@ -0,0 +1,217 @@
+}
+
+// Ready indicates if Manager can accept new config update requests
+func (mgr *Manager) Ready() bool {
+	if mgr.lock != 0 {
+		return false
+	}
+	return opamp.Ready()
+}
+
+func Initiate(db *sqlx.DB, engine string) error {
+	m.Repo = Repo{db}
+	return m.initDB(engine)
+}
+
+// Ready indicates if Manager can accept new config update requests
+func Ready() bool {
+	return m.Ready()
+}
+
+func GetLatestVersion(ctx context.Context, elementType ElementTypeDef) (*ConfigVersion, error) {
+	return m.GetLatestVersion(ctx, elementType)
+}
+
+func GetConfigVersion(ctx context.Context, elementType ElementTypeDef, version int) (*ConfigVersion, error) {
+	return m.GetConfigVersion(ctx, elementType, version)
+}
+
+func GetConfigHistory(ctx context.Context, typ ElementTypeDef) ([]ConfigVersion, error) {
+	return m.GetConfigHistory(ctx, typ)
+}
+
+// StartNewVersion launches a new config version for given set of elements
+func StartNewVersion(ctx context.Context, eleType ElementTypeDef, elementIds []string) (*ConfigVersion, error) {
+
+	if !m.Ready() {
+		// agent is already being updated, ask caller to wait and re-try after sometime
+		return nil, fmt.Errorf("agent updater is busy")
+	}
+
+	// create a new version
+	cfg := NewConfigversion(eleType)
+
+	// insert new config and elements into database
+	err := m.insertConfig(ctx, cfg, elementIds)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func Redeploy(ctx context.Context, typ ElementTypeDef, version int) error {
+	if !atomic.CompareAndSwapUint32(&m.lock, 0, 1) {
+		return fmt.Errorf("agent updater is busy")
+	}
+
+	defer atomic.StoreUint32(&m.lock, 0)
+
+	configVersion, err := GetConfigVersion(ctx, typ, version)
+	if err != nil {
+		zap.S().Debugf("failed to fetch config version during redeploy", err)
+		return fmt.Errorf("failed to fetch details of the config version")
+	}
+
+	if configVersion == nil || (configVersion != nil && configVersion.LastConf == "") {
+		zap.S().Debugf("config version has no conf yaml", configVersion)
+		return fmt.Errorf("the config version can not be redeployed")
+	}
+	switch typ {
+	case ElementTypeSamplingRules:
+		var config *tsp.Config
+		if err := yaml.Unmarshal([]byte(configVersion.LastConf), &config); err != nil {
+			zap.S().Errorf("failed to read last conf correctly", err)
+			return fmt.Errorf("failed to read the stored config correctly")
+		}
+
+		// merge current config with new filter params
+		processorConf := map[string]interface{}{
+			"signoz_tail_sampling": config,
+		}
+
+		opamp.AddToTracePipelineSpec("signoz_tail_sampling")
+		configHash, err := opamp.UpsertControlProcessors(ctx, "traces", processorConf, m.OnConfigUpdate)
+		if err != nil {
+			zap.S().Errorf("failed to call agent config update for trace processor:", err)
+			return fmt.Errorf("failed to deploy the config")
+		}
+
+		m.updateDeployStatus(ctx, ElementTypeSamplingRules, version, string(DeployInitiated), "Deployment started", configHash, configVersion.LastConf)
```

Comment:
The lastConf would be used to redeploy an older config.  This would be particularly useful when a new agent (fresh config) collector joins the ring.  We can pull latest configs from all 3 element types, merge and deploy.  

hence, we keep record of what was pushed. This reduces the effort of having to rebuild config and also cycling-dependency of agentConf on calling packages (e.g. ingestionRules)

Reply:
The agent config versions table stores versioned configs for each types - Logs, Sampling Rules and Drop rules. 

Together [agent_config_versions, agent_config_elements, log pipelines or rules] make up signoz definition of config.  These are used to prepare part of a collector config when the user is trying to push changes. 
 
The prepared collector part is stored in agent_config_versions table for each version. So in case we want to redeploy we don't fetch log pipelines and rules to re-make the collector config.

Reply:
>The agent config versions table stores versioned configs for each types - Logs, Sampling Rules and Drop rules.

Yes, this was clear. I understood this.

>Together [agent_config_versions, agent_config_elements, log pipelines or rules] make up signoz definition of config. These are used to prepare part of a collector config when the user is trying to push changes.

Where is the "signoz definition of config" defined?

> So in case we want to redeploy we don't fetch log pipelines and rules to re-make the collector config.

What is `Redeploy` supposed to do? Does it push the config for an element type `typ` with version `version` to the collector? If the idea is to fetch config from DB and push, why does it call `opamp.AddToTracePipelineSpec`?


> This would be particularly useful when a new agent (fresh config) collector joins the ring. We can pull latest configs from all 3 element types, merge and deploy.

The `UpsertControlProcessors` goes over all agents and pushes the changes. When a new agent joins we know if it is newly created. Why update all the agents instead of a newly joined agent?